### PR TITLE
Use TCP/IP to increase bandwidth and solve packet loss/ordering issues

### PR DESCRIPTION
**Pull request by bparks13**
Updated to match the logic in OpenEphys.Bonsai.EphysSocket, which allows for packets larger than the max UDP packet size by breaking up into smaller packets before sending.

### CODE REVIEW

File: Source/EphysSocket.cpp
Comment (jonnew) [inline Source/EphysSocket.cpp:220]: If statements without brackets are ok for things that would reasonably fit as a ternary operator but when there is no assignment. These statements are too long for that. Please use squigglies.

Comment (jonnew) [reply Source/EphysSocket.cpp:220]: `std::vector::insert` is for prepending an array dynamically. I'm not sure that's what you want. I think you have pre-allocated these arrays so what you really want is `std::vector:;assign`

File: Source/EphysSocket.cpp
Comment (jonnew) [inline Source/EphysSocket.cpp:260]: Why are you clearing the buffer? It will be overwritten on the next pass. No need.

File: Source/EphysSocket.cpp
Comment (jonnew) [reply Source/EphysSocket.cpp:243]: Additionally, again, why use `insert` when you could pre-allocate the buffer and then use `assign`.

File: Source/EphysSocket.cpp
Comment (jonnew) [reply Source/EphysSocket.cpp:220]: This also caused a hard crash when i set the buffer size parameters incorrectly. Make sure you add nonsense values into the GUI to ensure that exceptions are properly handled.

Comment (jonnew) [reply Source/EphysSocket.cpp:220]: Given all this might be worth putting the size of the matrix into the header. We are implicitly assuming some size information by sending the index, so its no less clean.

**Review by jonnew — CHANGES_REQUESTED** [review]

File: Source/EphysSocket.cpp
Comment (bparks13) [reply Source/EphysSocket.cpp:243]: You are correct, using `at` is not needed and has been removed.

Regarding `assign`, my understanding is that using it removes all existing data and adds in only the data given. To use it properly we would need to have converted all of the data (i.e., scale / offset / type conversion) before we could use it to fill the variable in one call. 

Additionally, we cannot place the data in as-is, since currently it is contiguous by channels; if the data is interleaved we would be a step closer to this though. I can transpose the data on the Bonsai side if that would help here.

File: Source/EphysSocket.cpp
Comment (bparks13) [reply Source/EphysSocket.cpp:220]: Currently I don't pre-allocate the data itself, only the size of the container. I am using `std::vector::insert` because there might be multiple packets, and using `std::vector::assign` would remove any previously added data elements. I couldn't find a method in `std::vector` that allows for overwriting particular blocks of the container, but I may have missed it.

Comment (bparks13) [reply Source/EphysSocket.cpp:220]: `std::copy` overwrites existing data in a vector, this is implemented now.

File: Source/EphysSocket.cpp
Comment (jonnew) [inline Source/EphysSocket.cpp:361]: Is this a remote procedure call handler? It looks like its trying to set offset and scale parameters which we removed in this refactor, right?

File: Source/EphysSocketEditor.cpp
Comment (jonnew) [inline Source/EphysSocketEditor.cpp:45]: Why make this type of aesthetic change? I think the current all caps was probably just keeping in line with other looks and feel conventions in the GUI

Comment (jonnew) [reply Source/EphysSocketEditor.cpp:45]: Now that i built and am looking at it, I understand. Looks good in the context where its needed. However, in the latest revision all this data is in the header and these options should be removed from the GUI

**Comment by anjaldoshi** [comment]
Hi, 

Looks like the plugin is failing to build on Linux (Ubuntu) and MacOS. 
https://github.com/open-ephys-plugins/ephys-socket/actions/runs/8271166810
https://github.com/open-ephys-plugins/ephys-socket/actions/runs/8271166809

This situation, combined with an incorrect merge to the main branch, has caused upstream problems in the cloud repository where the plugins are hosted. Specifically, the plugin files from the last version (0.2.1) were overwritten by the new version (0.3.0), resulting in no data being generated for the 0.3.0 version.

Before merging to main, you need to ensure a couple of things:
1. Make sure the plugin builds and runs successfully on all platforms supported by the plugin
2. Tag the latest commit with the new version number provided in [`OpenEphysLib.cpp`](https://github.com/open-ephys-plugins/ephys-socket/blob/main/Source/OpenEphysLib.cpp#L41). This step is crucial because the GitHub Actions script uses the latest tag number as the version when deploying to the cloud.

Now, since the plugin currently built successfully only on Windows, the files for version 0.2.1 of Ephys Socket for Windows were inadvertently overwritten with 0.3.0 binaries. Meanwhile, the Linux and MacOS binaries for v0.2.1 remain the actual v0.2.1 binaries.

**Comment by jonnew** [comment]
Sounds good. Brandon, I think you have push access to this repository
right? If not, I think you should so that you can fix stuff like this
directly

**Comment by jonnew** [comment]
Another note -- should there be some checks for successful builds in the
actions before pushing packages to a live server?

**Comment by anjaldoshi** [comment]
The action running for a platform doesn't check whether actions for other platforms failed or not before deploying.
One thing that we can do, and is worth doing, is to run the actions for `Pull Requests`. Currently, it only runs on any commit/tag push. Since Brandon's fork didn't have Actions enabled, it didn't run for him when he pushed changes to his fork. In such cases, having actions run for `Pull Requests` and allowing PRs to be merged only on successful checks for all actions is the best approach.

**Pull request merged**
